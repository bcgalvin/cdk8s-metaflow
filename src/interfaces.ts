import { HttpIngressPathType, ImagePullPolicy, ServiceType } from './enums';
import { Affinity, EnvFromSource, IngressTls, SecurityContext, Toleration } from './imports/k8s';

export interface ImageOptions {
  readonly repository?: string;
  readonly pullPolicy?: ImagePullPolicy;
  readonly tag?: string;
}

export interface ServiceAccountOptions {
  readonly create?: boolean | undefined;
  readonly annotations?: Record<string, string>;
  readonly name?: string;
}

export interface DbMigrationOptions {
  readonly runOnStart?: boolean;
  readonly onlyIfDbEmpty?: boolean;
}

export interface ServiceOptions {
  readonly type?: ServiceType;
  readonly port?: number;
  readonly annotations?: Record<string, string>;
}

export interface HttpIngressPath {
  readonly path?: string;
  readonly pathType?: HttpIngressPathType;
}

export interface HostOptions {
  readonly host?: string;
  readonly paths?: HttpIngressPath[];
}

export interface IngressOptions {
  readonly enabled?: boolean;
  readonly className?: string;
  readonly annotations?: Record<string, string>;
  readonly hosts?: HostOptions[];
  readonly tls?: IngressTls;
}

export interface AutoscalingOptions {
  readonly enabled?: boolean;
  readonly minReplicas?: number;
  readonly maxReplicas?: number;
  readonly targetCPUUtilizationPercentage?: number;
}

export interface IApiResource {
  readonly apiGroup?: string;
  readonly resourceType?: string;
  readonly resourceName?: string;
}

export interface MetadataDatabaseEnvOptions {
  readonly port?: number;
  readonly password?: string;
  readonly host?: string;
  readonly user?: string;
}

export interface MetaflowServiceOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly dbMigrations?: DbMigrationOptions;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
}

export interface MetaflowUIOptions {
  readonly replicaCount?: number;
  readonly image?: ImageOptions;
  readonly uiImage?: ImageOptions;
  readonly imagePullSecrets?: string[];
  readonly nameOverride?: string;
  readonly fullnameOverride?: string;
  readonly env: Record<string, string>;
  readonly serviceAccount?: ServiceAccountOptions;
  readonly podAnnotations?: Record<string, string>;
  readonly podSecurityContext?: Record<string, string>;
  readonly securityContext?: SecurityContext;
  readonly service?: ServiceOptions;
  readonly serviceStatic?: ServiceOptions;
  readonly ingress?: IngressOptions;
  readonly resources?: IApiResource[];
  readonly autoscaling?: AutoscalingOptions;
  readonly nodeSelector?: Record<string, string>;
  readonly tolerations?: Toleration[];
  readonly affinity?: Affinity;
  readonly envFrom?: EnvFromSource[];
  readonly metadatadb?: MetadataDatabaseEnvOptions;
  readonly metaflowDatastoreSysrootS3?: string;
}

export interface DatabaseAuthOptions {
  readonly database?: string;
  readonly password?: string;
  readonly username?: string;
  readonly postgresPassword?: string;
  readonly enablePostgresUser?: boolean;
  readonly replicationPassword?: string;
  readonly replicationUsername?: string;
}

export interface DatabaseMetricsOptions {
  readonly enabled?: boolean;
}

export interface DatabaseResourceRequestOptions {
  readonly memory?: string;
  readonly cpu?: string;
}

export interface DatabaseResourcesOptions {
  readonly requests?: DatabaseResourceRequestOptions;
}

export interface DatabaseReplicationOptions {
  readonly enabled?: boolean;
  readonly readReplicas?: number;
}

export interface DatabaseVolumePermissionsOptions {
  readonly enabled?: boolean;
}

export interface MetadataDatabaseOptions {
  readonly auth?: DatabaseAuthOptions;
  readonly metrics?: DatabaseMetricsOptions;
  readonly resources?: DatabaseResourcesOptions;
  readonly replication?: DatabaseReplicationOptions;
  readonly architecture?: string;
  readonly volumePermissions?: DatabaseVolumePermissionsOptions;
}
