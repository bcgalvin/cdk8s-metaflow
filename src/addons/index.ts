export * from './addon';
export * from './postgres';
export * from './argo-workflows';
export * from './autoscaler';
