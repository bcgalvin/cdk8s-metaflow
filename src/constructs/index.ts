export * from './metaflow-service';
export * from './interfaces';
