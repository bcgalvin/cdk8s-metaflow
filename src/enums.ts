export enum ImagePullPolicy {
  ALWAYS = 'Always',
  IF_NOT_PRESENT = 'IfNotPresent',
  NEVER = 'Never',
}

export enum HttpIngressPathType {
  PREFIX = 'Prefix',
  EXACT = 'Exact',
  IMPLEMENTATION_SPECIFIC = 'ImplementationSpecific',
}

export enum ServiceType {
  CLUSTER_IP = 'ClusterIP',
  NODE_PORT = 'NodePort',
  LOAD_BALANCER = 'LoadBalancer',
  EXTERNAL_NAME = 'ExternalName',
}
