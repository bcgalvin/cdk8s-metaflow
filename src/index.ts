export * from './constructs';
export * from './addons';
