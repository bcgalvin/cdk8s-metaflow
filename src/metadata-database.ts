import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetadataDatabaseOptions } from './interfaces';

export interface MetadataDatabaseProps {
  readonly chartVersion: string;
  readonly chartValues?: MetadataDatabaseOptions;
}

export class MetadataDatabase extends Construct {
  constructor(scope: Construct, id: string, props: MetadataDatabaseProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {
      enabled: true,
      auth: {
        database: 'metaflow',
        password: 'metaflow',
        username: 'metaflow',
      },
    };

    new Helm(this, 'postgres', {
      chart: 'bitnami/postgresql',
      releaseName: 'release-name-postgresql',
      helmFlags: ['--version', props.chartVersion],
      values: chartValues,
    });
  }
}
