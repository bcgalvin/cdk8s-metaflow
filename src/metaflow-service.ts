import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetaflowServiceOptions } from './interfaces';

export interface MetaflowServiceProps {
  readonly chartVersion: string;
  readonly chartValues?: MetaflowServiceOptions;
}

export class MetaflowService extends Construct {
  constructor(scope: Construct, id: string, props: MetaflowServiceProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {};

    new Helm(this, 'metaflow-service-helm', {
      chart: 'metaflow/metaflow-service',
      releaseName: 'release-name-metaflow-service',
      helmFlags: ['--version', props.chartVersion],
      values: { ...chartValues },
    });
  }
}
