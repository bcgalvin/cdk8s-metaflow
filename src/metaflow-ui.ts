import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { MetaflowUIOptions } from './interfaces';

export interface MetaflowUIProps {
  readonly chartVersion: string;
  readonly chartValues?: MetaflowUIOptions;
}

export class MetaflowUI extends Construct {
  constructor(scope: Construct, id: string, props: MetaflowUIProps) {
    super(scope, id);

    const chartValues = props.chartValues ?? {};

    new Helm(this, 'metaflow-ui-helm', {
      chart: 'metaflow/metaflow-ui',
      releaseName: 'release-name-metaflow-ui',
      helmFlags: ['--version', props.chartVersion],
      values: chartValues,
    });
  }
}
