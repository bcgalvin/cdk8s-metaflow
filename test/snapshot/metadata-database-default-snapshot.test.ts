import { MetadataDatabase } from '../../src';
// @ts-ignore
import { synth } from '../util';

describe('metadata database', () => {
  test('default snapshot', () => {
    // GIVEN
    const chart = synth(MetadataDatabase, {
      chartVersion: '10.16.2',
    });
    // THEN
    expect(chart).toMatchSnapshot();
  });
});
