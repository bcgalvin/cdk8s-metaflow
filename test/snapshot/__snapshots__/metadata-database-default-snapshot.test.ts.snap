// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata database default snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "data": Object {
      "postgresql-password": "Wk9LajN1NklUWA==",
    },
    "kind": "Secret",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-postgresql",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "postgresql",
        "helm.sh/chart": "postgresql-10.16.2",
      },
      "name": "release-name-postgresql",
      "namespace": "default",
    },
    "type": "Opaque",
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-postgresql",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "postgresql",
        "helm.sh/chart": "postgresql-10.16.2",
        "service.alpha.kubernetes.io/tolerate-unready-endpoints": "true",
      },
      "name": "release-name-postgresql-headless",
      "namespace": "default",
    },
    "spec": Object {
      "clusterIP": "None",
      "ports": Array [
        Object {
          "name": "tcp-postgresql",
          "port": 5432,
          "targetPort": "tcp-postgresql",
        },
      ],
      "publishNotReadyAddresses": true,
      "selector": Object {
        "app.kubernetes.io/instance": "release-name-postgresql",
        "app.kubernetes.io/name": "postgresql",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/instance": "release-name-postgresql",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "postgresql",
        "helm.sh/chart": "postgresql-10.16.2",
      },
      "name": "release-name-postgresql",
      "namespace": "default",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "name": "tcp-postgresql",
          "port": 5432,
          "targetPort": "tcp-postgresql",
        },
      ],
      "selector": Object {
        "app.kubernetes.io/instance": "release-name-postgresql",
        "app.kubernetes.io/name": "postgresql",
        "role": "primary",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": Object {
      "labels": Object {
        "app.kubernetes.io/component": "primary",
        "app.kubernetes.io/instance": "release-name-postgresql",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "postgresql",
        "helm.sh/chart": "postgresql-10.16.2",
      },
      "name": "release-name-postgresql-postgresql",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "app.kubernetes.io/instance": "release-name-postgresql",
          "app.kubernetes.io/name": "postgresql",
          "role": "primary",
        },
      },
      "serviceName": "release-name-postgresql-headless",
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app.kubernetes.io/component": "primary",
            "app.kubernetes.io/instance": "release-name-postgresql",
            "app.kubernetes.io/managed-by": "Helm",
            "app.kubernetes.io/name": "postgresql",
            "helm.sh/chart": "postgresql-10.16.2",
            "role": "primary",
          },
          "name": "release-name-postgresql",
        },
        "spec": Object {
          "affinity": Object {
            "podAntiAffinity": Object {
              "preferredDuringSchedulingIgnoredDuringExecution": Array [
                Object {
                  "podAffinityTerm": Object {
                    "labelSelector": Object {
                      "matchLabels": Object {
                        "app.kubernetes.io/component": "primary",
                        "app.kubernetes.io/instance": "release-name-postgresql",
                        "app.kubernetes.io/name": "postgresql",
                      },
                    },
                    "namespaces": Array [
                      "default",
                    ],
                    "topologyKey": "kubernetes.io/hostname",
                  },
                  "weight": 1,
                },
              ],
            },
          },
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "BITNAMI_DEBUG",
                  "value": "false",
                },
                Object {
                  "name": "POSTGRESQL_PORT_NUMBER",
                  "value": "5432",
                },
                Object {
                  "name": "POSTGRESQL_VOLUME_DIR",
                  "value": "/bitnami/postgresql",
                },
                Object {
                  "name": "PGDATA",
                  "value": "/bitnami/postgresql/data",
                },
                Object {
                  "name": "POSTGRES_USER",
                  "value": "postgres",
                },
                Object {
                  "name": "POSTGRES_PASSWORD",
                  "valueFrom": Object {
                    "secretKeyRef": Object {
                      "key": "postgresql-password",
                      "name": "release-name-postgresql",
                    },
                  },
                },
                Object {
                  "name": "POSTGRESQL_ENABLE_LDAP",
                  "value": "no",
                },
                Object {
                  "name": "POSTGRESQL_ENABLE_TLS",
                  "value": "no",
                },
                Object {
                  "name": "POSTGRESQL_LOG_HOSTNAME",
                  "value": "false",
                },
                Object {
                  "name": "POSTGRESQL_LOG_CONNECTIONS",
                  "value": "false",
                },
                Object {
                  "name": "POSTGRESQL_LOG_DISCONNECTIONS",
                  "value": "false",
                },
                Object {
                  "name": "POSTGRESQL_PGAUDIT_LOG_CATALOG",
                  "value": "off",
                },
                Object {
                  "name": "POSTGRESQL_CLIENT_MIN_MESSAGES",
                  "value": "error",
                },
                Object {
                  "name": "POSTGRESQL_SHARED_PRELOAD_LIBRARIES",
                  "value": "pgaudit",
                },
              ],
              "image": "docker.io/bitnami/postgresql:11.14.0-debian-10-r28",
              "imagePullPolicy": "IfNotPresent",
              "livenessProbe": Object {
                "exec": Object {
                  "command": Array [
                    "/bin/sh",
                    "-c",
                    "exec pg_isready -U \\"postgres\\" -h 127.0.0.1 -p 5432",
                  ],
                },
                "failureThreshold": 6,
                "initialDelaySeconds": 30,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "name": "release-name-postgresql",
              "ports": Array [
                Object {
                  "containerPort": 5432,
                  "name": "tcp-postgresql",
                },
              ],
              "readinessProbe": Object {
                "exec": Object {
                  "command": Array [
                    "/bin/sh",
                    "-c",
                    "-e",
                    "exec pg_isready -U \\"postgres\\" -h 127.0.0.1 -p 5432
[ -f /opt/bitnami/postgresql/tmp/.initialized ] || [ -f /bitnami/postgresql/.initialized ]
",
                  ],
                },
                "failureThreshold": 6,
                "initialDelaySeconds": 5,
                "periodSeconds": 10,
                "successThreshold": 1,
                "timeoutSeconds": 5,
              },
              "resources": Object {
                "requests": Object {
                  "cpu": "250m",
                  "memory": "256Mi",
                },
              },
              "securityContext": Object {
                "runAsUser": 1001,
              },
              "volumeMounts": Array [
                Object {
                  "mountPath": "/dev/shm",
                  "name": "dshm",
                },
                Object {
                  "mountPath": "/bitnami/postgresql",
                  "name": "data",
                },
              ],
            },
          ],
          "securityContext": Object {
            "fsGroup": 1001,
          },
          "volumes": Array [
            Object {
              "emptyDir": Object {
                "medium": "Memory",
              },
              "name": "dshm",
            },
          ],
        },
      },
      "updateStrategy": Object {
        "type": "RollingUpdate",
      },
      "volumeClaimTemplates": Array [
        Object {
          "metadata": Object {
            "name": "data",
          },
          "spec": Object {
            "accessModes": Array [
              "ReadWriteOnce",
            ],
            "resources": Object {
              "requests": Object {
                "storage": "8Gi",
              },
            },
          },
        },
      ],
    },
  },
]
`;
